### PR TITLE
feat(openchallenges): add more unit tests to the image service exception to increase coverage to > 50%

### PR DESCRIPTION
**Pull request by mdsage1**
## Description

Added additional unit tests for files that were not autogenerated by OpenAPI. Investigated several options for auto generating unit tests forOpen API auto generated files. None of the options I found seemed helpful and many were associated w/Open API competitors.

## Related Issue

feat(openchallenges): added unit tests to the image service [#2291](https://github.com/Sage-Bionetworks/sage-monorepo/pull/2291)
Add unit test for getting a single organization [#1528](https://github.com/Sage-Bionetworks/sage-monorepo/issues/1528)
[Story] Write a unit test for the enpoint of the image service [#1456](https://github.com/Sage-Bionetworks/sage-monorepo/issues/1456)
feat: Create unit test for the image service [#1472](https://github.com/Sage-Bionetworks/sage-monorepo/pull/1472)
Fixes #(issue)

## Changelog
 Added additional unit tests that increased coverage to 59%:
image

- Add
1. apps/openchallenges/image-service/src/test/java/org/sagebionetworks/exception/BadRequestExceptionTest.java
2. apps/openchallenges/image-service/src/test/java/org/sagebionetworks/exception/ControllerAdvisorTest.java
3. apps/openchallenges/image-service/src/test/java/org/sagebionetworks/exception/ErrorConstantsTest.java
4. apps/openchallenges/image-service/src/test/java/org/sagebionetworks/openchallenges/image/service/exception/SimpleChallengeGlobalExceptionTest.java
-Update

1. apps/openchallenges/image-service/src/test/java/org/sagebionetworks/openchallenges/image/service/service/ImageServiceTest.java

### CODE REVIEW

**Comment by mdsage1** [comment]
@tschaffter Can you please take a look at the additional unit tests I added for image service last week?

**Comment by tschaffter** [comment]
> Investigated several options for auto generating unit tests forOpen API auto generated files. None of the options I found seemed helpful and many were associated w/Open API competitors.

What did you investigate exactly? What options did you found?

File: apps/openchallenges/image-service/src/test/java/org/sagebionetworks/exception/BadRequestExceptionTest.java
Comment (tschaffter) [inline apps/openchallenges/image-service/src/test/java/org/sagebionetworks/exception/BadRequestExceptionTest.java:1]: The location of this file should mirror the folder structure of `src/main`. Here, `apps/openchallenges/image-service/src/test/java/org/sagebionetworks/exception/BadRequestExceptionTest.java` should be `apps/openchallenges/image-service/src/test/java/org/sagebionetworks/openchallenges/challenge/service/exception/BadRequestExceptionTest.java`.

The same comment applies to the other files and also files submitted in the previous PR.

File: apps/openchallenges/image-service/src/test/java/org/sagebionetworks/exception/BadRequestExceptionTest.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package org.sagebionetworks.openchallenges.image.service.exception;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.junit.jupiter.api.Test;
+
+public class BadRequestExceptionTest {
+
+  @Test
+  public void BadRequestException_ConstructorTypeShouldMatch() {
```
Comment (tschaffter) [inline apps/openchallenges/image-service/src/test/java/org/sagebionetworks/exception/BadRequestExceptionTest.java:10]: Use consistent function name format. This format is different from the first tests you have created, e.g. `getImage_ShouldReturnImage_WhenObjectKeyIsPassed()`. 

Also, function names should always start with a lowercase.

The same comments applies to other functions that are part of this PR.

**Review by tschaffter — CHANGES_REQUESTED** [review]

**Comment by mdsage1** [comment]
> > Investigated several options for auto generating unit tests forOpen API auto generated files. None of the options I found seemed helpful and many were associated w/Open API competitors.
> 
> What did you investigate exactly? What options did you found?

1. I watched a YouTube video and found this PP [How_to_fast_generate_your_API_Test_with_OpenAPI_Tools_and_Rest-Assured.pdf](https://github.com/Sage-Bionetworks/sage-monorepo/files/13271602/How_to_fast_generate_your_API_Test_with_OpenAPI_Tools_and_Rest-Assured.pdf). This one didn't really help me to understand the "How" of implementing automatic test generation for auto generated modules they way I hoped it would.
2. This [link](https://www.merge.dev/blog/openapi-tutorial-how-to-automatically-generate-tests-for-openapi-generator-sdks) seemed to address the autogeneration using a .mustache file (which I could not find an example of in our file structure). It could be added but I have no idea how to determine if classes are optionally included but, there is this [template](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/python/model_test.mustache) that may be helpful if this method is used. 

I also am unsure where in the process this would be implemented in order to use it as described by the author below:

> You can download that file to your local directory, and use it in the docker SDK generation command from the previous section [by adding this flag](https://openapi-generator.tech/docs/templating/):
> `-t ./local/REPLACE-WITH-YOUR-TEMPLATES-FOLDER$`

Here are some other tools I found that seemed like they would not suit our needs:

- [newman](https://learning.postman.com/docs/collections/using-newman-cli/command-line-integration-with-newman/)
- [Humlix](https://www.humlix.com/#howitworks)

File: apps/openchallenges/image-service/src/test/java/org/sagebionetworks/exception/BadRequestExceptionTest.java
Comment (mdsage1) [reply apps/openchallenges/image-service/src/test/java/org/sagebionetworks/exception/BadRequestExceptionTest.java:1]: Moved the files to the correct directory.

**Comment by mdsage1** [comment]
Updated the test names in all new tests for image-service. Restructured the file order to match that specified by @tschaffter

**Comment by tschaffter** [comment]
@mdsage1 Thanks for sharing the resources. REST Assured describes an interesting framework that could make it easier to develop tests. However, this approach does not seem to apply to develop test for internal and other utility functions.

The approach that use templates could be used to write tests for the code generated by OpenAPI generator. I know that such templates do exist but it depends on the generator being used. For example, generating a REST API for Python comes with automatically generated tests. I'm not sure whether we could get the benefit for the generator we use for Java. I will have a look.

One topic we can discuss together offline is the format of the name of the test functions. The format you are using is described in the article below from Microsoft. Note that in this context, you were right to capitalize the first letter of the test. It seems that this naming convention for test takes precedence over the more general rule that function should start with a lowercase. So feel free to capitalize again the first letter.

https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices

See [Google Java naming convention](https://google.github.io/styleguide/javaguide.html#s5.2.3-method-names) about function names. They do use camel case for test

For now please continue with the naming convention you have been applying so far.

**Comment by tschaffter** [comment]
The coverage of the package `exception` could still be increased. Is this something that you could do as part of this PR? I propose to write all the tests required to reach 50% coverage in this PR instead of creating multiple PRs. Alternatively, it would be OK to add PR package by package and open a PR for each of them. What is important is that PRs represent a logical chunk of work (e.g. "add unit tests to the package `exception` of the image service" or "increase the image service coverage to 50%").

**Comment by tschaffter** [comment]
> **Note**
> Feel free to request a new review from a reviewer by clicking on the "reload" button next to the reviewer's name.

**Comment by tschaffter** [comment]
@mdsage1 This is where the mustache template files for the generator `spring` live. Note that this folder includes the latest templates files while we may be using a different version (defined in our file `openapitools.json`).

https://github.com/OpenAPITools/openapi-generator/tree/master/modules/openapi-generator/src/main/resources/JavaSpring

 I can confirm that the generator `spring` does not come with templates for test files with one exception. The template below creates a single, virtually empty test file that is part of our code base.

https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/JavaSpring/libraries/spring-boot/SpringBootTest.mustache

**Comment by tschaffter** [comment]
Generated code will be excluded from test coverage with #2332

**Comment by tschaffter** [comment]
@mdsage1 When working on a PR, e.g. when addressing feedback, please convert the PR back to Draft using the link `Convert to Draft` at the top-right corner of the page.

**Comment by mdsage1** [comment]
image-service unit test coverage has been improved to 59% after addingan adiitonal test to exception for controller advisor. Reformatted naming as requested.

**Comment by mdsage1** [comment]
> Note Feel free to request a new review from a reviewer by clicking on the "reload" button next to the reviewer's name (if they have already left a review).

@tschaffter I don't see a reload button so I'm tagging you. Maybe we can look at this in our meeting for the next PR.

**Comment by tschaffter** [comment]
@mdsage1 What has changed since the last review you requested? Please update the original post in this thread, in particular the Preview section.

**Comment by mdsage1** [comment]
Add

1. apps/openchallenges/image-service/src/test/java/org/sagebionetworks/openchallenges/image/service/exception/ControllerAdvisorTest.java --added additional tests
2. apps/openchallenges/image-service/src/test/java/org/sagebionetworks/openchallenges/image/service/exception/BadRequestExceptionTest.java

Removed

1. apps/openchallenges/image-service/src/test/java/org/sagebionetworks/openchallenges/image/service/api/ImageApiDelegateImplTest.java (couldn't get it to work)

**Review by tschaffter — APPROVED** [review]

**Pull request merged**